// zotero-plugin/create-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Elements = exports.NAMESPACE = void 0;
exports.NAMESPACE = {
    XUL: 'http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul',
    HTML: 'http://www.w3.org/1999/xhtml',
};
class Elements {
    prefix;
    document;
    static all = [];
    static removeAll() {
        for (const ref of this.all) {
            try {
                const elements = ref.deref();
                if (elements)
                    elements.document.querySelectorAll(`.${elements.className}`).forEach(e => e.remove());
            }
            catch { }
        }
        this.all = [];
    }
    className;
    constructor(prefix, document) {
        this.prefix = prefix;
        this.document = document;
        this.className = `${prefix}-${Math.random()}-${Date.now()}`.replace('.', '-');
        if (this.document.createXULElement)
            Elements.all.push(new WeakRef(this));
    }
    serialize(node) {
        const s = new XMLSerializer();
        return s.serializeToString(node);
    }
    create(name, attrs = {}) {
        const children = attrs.$ || [];
        delete attrs.$;
        const namespace = name.startsWith('html:') ? exports.NAMESPACE.HTML : exports.NAMESPACE.XUL;
        name = name.replace('html:', '');
        const elt = this.document.createXULElement
            ? this.document[namespace === exports.NAMESPACE.XUL ? 'createXULElement' : 'createElement'](name)
            : this.document.createElementNS(namespace, name);
        attrs.class = `${this.className} ${attrs.class || ''}`.trim();
        for (const [a, v] of Object.entries(attrs)) {
            if (typeof v === 'string') {
                elt.setAttribute(a, v);
            }
            else if (typeof v === 'number') {
                elt.setAttribute(a, `${v}`);
            }
            else if (a.startsWith('on') && typeof v === 'function') {
                elt.addEventListener(a.replace('on', ''), event => {
                    ;
                    v(event)?.catch?.(err => {
                        throw err;
                    });
                });
            }
            else {
                throw new Error(`unexpected attribute ${a} of type ${typeof v}`);
            }
        }
        for (const child of children) {
            elt.appendChild(child);
        }
        return elt;
    }
    remove() {
        for (const elt of Array.from(this.document.getElementsByClassName(this.className))) {
            elt.remove();
        }
    }
}
exports.Elements = Elements;
