// zotero-plugin/logger.js
"use strict";
/* eslint-disable @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-empty-function, no-restricted-syntax */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
function stringifyXPCOM(obj) {
    if (!obj.QueryInterface)
        return '';
    if (obj.message)
        return `[XPCOM error ${obj.message}]`;
    if (obj.name)
        return `[XPCOM object ${obj.name}]`;
    return '[XPCOM object]';
}
function stringifyError(obj) {
    if (obj instanceof Error)
        return `[error: ${obj.message || '<unspecified error>'}\n${obj.stack}]`;
    // guess it is an errorevent
    if (obj.error instanceof Error && obj.message)
        return `[errorevent: ${obj.message} ${stringifyError(obj.error)}]`;
    if (typeof ErrorEvent !== 'undefined' && obj instanceof ErrorEvent)
        return `[errorevent: ${obj.message || '<unspecified errorevent>'}]`;
    return '';
}
function replacer() {
    const seen = new WeakSet();
    return (key, value) => {
        if (typeof value === 'object' && value !== null) {
            if (seen.has(value))
                return '[Circular]';
            seen.add(value);
        }
        if (value === null)
            return value;
        if (value instanceof Set)
            return [...value];
        if (value instanceof Map)
            return Object.fromEntries(value);
        switch (typeof value) {
            case 'string':
            case 'number':
            case 'boolean':
                return value;
            case 'object':
                return stringifyXPCOM(value) || stringifyError(value) || value;
        }
        if (Array.isArray(value))
            return value;
        return undefined;
    };
}
class Logger {
    id;
    phase = '';
    indent;
    constructor(id) {
        this.id = id;
    }
    #prefix(error) {
        return `{${error ? 'error: ' : ''}${this.phase}${this.id}} `;
    }
    debug(...msg) {
        Zotero.debug(`${this.#prefix()}${this.format(...msg)}\n`);
    }
    info(...msg) {
        Zotero.debug(`${this.#prefix()}${this.format(...msg)}\n`);
    }
    error(...msg) {
        Zotero.debug(`${this.#prefix(true)}${this.format(...msg)}\n`);
    }
    dump(msg, error) {
        if (error) {
            dump(`${this.#prefix(error)}${this.format(msg, error)}\n`);
        }
        else {
            dump(`${this.#prefix()}${this.format(msg)}\n`);
        }
    }
    to_s(obj) {
        if (typeof obj === 'string')
            return obj;
        return JSON.stringify(obj, replacer(), this.indent);
    }
    format(...msg) {
        return msg.map(m => this.to_s(m)).join(' ');
    }
}
exports.Logger = Logger;
