// zotero-plugin/manifest.js
"use strict";
/* eslint-disable no-console, @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/restrict-template-expressions, no-magic-numbers */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const glob_1 = require("glob");
const path = tslib_1.__importStar(require("path"));
const properties_reader_1 = tslib_1.__importDefault(require("properties-reader"));
const uri_templates_1 = tslib_1.__importDefault(require("uri-templates"));
const root_1 = tslib_1.__importDefault(require("./root"));
const version_1 = tslib_1.__importDefault(require("./version"));
const pkg = { ...require(path.join(root_1.default, 'package.json')) };
if (!pkg.id)
    pkg.id = `${pkg.name}@${pkg.author.email.replace(/.*@/, '')}`.toLowerCase();
if (pkg.xpi)
    Object.assign(pkg, pkg.xpi);
pkg.version = version_1.default;
if (pkg.updateLink)
    pkg.updateLink = (0, uri_templates_1.default)(pkg.updateLink).fill({ version: pkg.version });
pkg.updateURL = `${pkg.xpi.releaseURL}update.rdf`;
const translations = (0, glob_1.globSync)(path.join(root_1.default, 'locale/*/*.properties'));
for (const translation of translations) {
    const locale = path.basename(path.dirname(translation));
    const properties = (0, properties_reader_1.default)(translation);
    const description = properties.get('xpi.description');
    if (!description)
        continue;
    if (locale === 'en-US') {
        pkg.description = description;
    }
    else {
        pkg.localizedDescriptions = pkg.localizedDescriptions || {};
        pkg.localizedDescriptions[locale] = description;
    }
}
const options_and_vars = { minVersion: '7.0.0', maxVersion: '8.*', ...pkg, pretty: true };
try {
    Object.assign(options_and_vars, JSON.parse(fs.readFileSync(path.join(root_1.default, 'schema', 'supported.json'), 'utf8')));
}
catch (err) { // eslint-disable-line @typescript-eslint/no-unused-vars
    // ignore
}
console.log('generating updates.json');
fs.writeFileSync(path.join(root_1.default, 'gen/updates.json'), JSON.stringify({
    addons: {
        [pkg.id]: {
            updates: [
                {
                    version: options_and_vars.version,
                    update_link: options_and_vars.updateLink,
                    applications: {
                        zotero: {
                            strict_min_version: options_and_vars.minVersion,
                            strict_max_version: options_and_vars.maxVersion,
                        },
                    },
                },
            ],
        },
    },
}, null, 2));
const icons = [
    { 48: pkg.xpi?.iconURL?.replace(/^chrome:\/\/[^/]+\//, '') },
].filter(i => i[48]);
const basename = pkg.id.replace(/@.*/, '');
for (const i of [`content/skin/${basename}.png`, `skin/${basename}.png`, `${basename}.png`, 'icon.png']) {
    icons.push({ 48: i });
    icons.push({ 48: i.replace('/zotero-', '/') });
}
for (const i of [...icons]) {
    icons.push({ 48: i[48].replace(/[.](svg|png)$/, ext => ({ '.svg': '.png', '.png': '.svg' }[ext])) });
}
for (const i of [...icons]) {
    if (i[48].endsWith('.svg')) {
        i[96] = i[48];
    }
    else {
        i[96] = i[48].replace(/([.][^.]+)$/, '@2x$1');
    }
}
const icon = icons.find(i => fs.existsSync(path.join(root_1.default, ...i[48].split('/'))));
if (icon) {
    options_and_vars.icons = {
        48: icon[48],
        96: fs.existsSync(path.join(root_1.default, ...icon[96].split('/'))) ? icon[96] : icon[48],
    };
}
console.log('generating manifest.json');
fs.writeFileSync(path.join(root_1.default, 'build/manifest.json'), JSON.stringify({
    manifest_version: 2,
    name: options_and_vars.name,
    version: options_and_vars.version,
    description: options_and_vars.description,
    icons: options_and_vars.icons,
    applications: {
        zotero: {
            id: options_and_vars.id,
            update_url: options_and_vars.updateURL.replace('/update.rdf', '/updates.json'),
            strict_min_version: options_and_vars.minVersion,
            strict_max_version: options_and_vars.maxVersion,
        },
    },
}, null, 2));
